arenadata_toolkit: add TOAST's size to adb_relation_storage_size

red1452 opened the pull request:
This patch adds TOAST table's size to summary size of relation at C function
`adb_relation_storage_size`, which is used at SQL functions
`adb_relation_storage_size` and `adb_relation_storage_size_on_segments`.

The code is based on `calculate_toast_table_size` from dbsize.c, but without
calculating size of TOAST's indexes.

New test `adb_relation_storage_size_test` was added to check calculating sizes of
different tables (heap, ao, with TOAST or without) at SQL functions
`adb_relation_storage_size` and `adb_relation_storage_size_on_segments`.

red1452 committed:
arenadata_toolkit: add toast's size to adb_relation_storage_size

This patch adds toast table's size to summary size of relation at C function
adb_relation_storage_size, which is used at SQL functions adb_relation_storage_size and
adb_relation_storage_size_on_segments.

The code is based on calculate_toast_table_size from dbsize.c, but without calculating
size of toast's indexes.

New test adb_relation_storage_size_test was added to check calculating
sizes of different tables (heap, ao, with toast or without) at SQL functions
adb_relation_storage_size and adb_relation_storage_size_on_segments.

red1452 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4476

red1452 committed:
fix comment at test

BenderArenadata commented:
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/848384

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/848385

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/848386

BenderArenadata commented:
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/849052

RekGRpth commented on gpcontrib/arenadata_toolkit/sql/adb_relation_storage_size_test.sql:
```diff
@@ -0,0 +1,89 @@
+DISTRIBUTED BY (a);
+
+-- Check that toast exist only for "with_toast" tables
+SELECT relname
```
```suggestion
SELECT relname, reltoastrelid != 0 with_toast
FROM pg_class
WHERE relname IN ('heap_table_with_toast', 'heap_table_without_toast', 'ao_table_with_toast', 'ao_table_without_toast');
```

RekGRpth commented on gpcontrib/arenadata_toolkit/sql/adb_relation_storage_size_test.sql:
```diff
@@ -0,0 +1,89 @@
+INSERT INTO ao_table_without_toast SELECT i, i*10 FROM generate_series(1,15) AS i;
+
+-- Check sizes on segments
+SELECT * FROM arenadata_toolkit.adb_relation_storage_size_on_segments((
```
```suggestion
WITH s AS (
    SELECT oid, relname
    FROM pg_class
    WHERE relname IN ('heap_table_with_toast', 'heap_table_without_toast', 'ao_table_with_toast', 'ao_table_without_toast')
) SELECT relname, size FROM s, arenadata_toolkit.adb_relation_storage_size_on_segments(oid);
```

RekGRpth commented on gpcontrib/arenadata_toolkit/sql/adb_relation_storage_size_test.sql:
```diff
@@ -0,0 +1,89 @@
+    FROM generate_series(1,50000))
+WHERE a = 1;
+
+SELECT * FROM arenadata_toolkit.adb_relation_storage_size_on_segments((
```
```suggestion
WITH s AS (
    SELECT oid, relname
    FROM pg_class
    WHERE relname IN ('heap_table_with_toast', 'ao_table_with_toast')
) SELECT relname, size FROM s, arenadata_toolkit.adb_relation_storage_size_on_segments(oid);
```

RekGRpth commented on gpcontrib/arenadata_toolkit/sql/adb_relation_storage_size_test.sql:
```diff
@@ -0,0 +1,89 @@
+ORDER BY 1;
+
+-- Check summary size of tables
+SELECT * FROM arenadata_toolkit.adb_relation_storage_size((
```
```suggestion
WITH s AS (
    SELECT oid, relname
    FROM pg_class
    WHERE relname IN ('heap_table_with_toast', 'heap_table_without_toast', 'ao_table_with_toast', 'ao_table_without_toast')
) SELECT relname, adb_relation_storage_size size FROM s, arenadata_toolkit.adb_relation_storage_size(oid);
```

red1452 committed:
test's optimization

red1452 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4476

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/59893

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/858617

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/858608

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/858611

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/858610

red1452 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4476

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/59918

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/860229

red1452 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4476

red1452 committed:
arenadata_toolkit: add toast's size to adb_relation_storage_size (#654)

This patch adds toast table's size to summary size of relation at C function
adb_relation_storage_size, which is used at SQL functions
adb_relation_storage_size and adb_relation_storage_size_on_segments.

The code is based on calculate_toast_table_size from dbsize.c, but without
calculating size of toast's indexes.

New test adb_relation_storage_size_test was added to check calculating sizes of
different tables (heap, ao, with toast or without) at SQL functions
adb_relation_storage_size and adb_relation_storage_size_on_segments.

red1452 commented:
> @red1452 , Ready to approve. Format your patch description with proper 80 aligment, please.

Done.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60148

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/872166

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/872167

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/872164

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/872160

bimboterminator1 commented on gpcontrib/arenadata_toolkit/expected/adb_relation_storage_size_test.out:
```diff
@@ -0,0 +1,95 @@
+ORDER BY 1, 2;
+        relname        | gp_segment_id | size  
+-----------------------+---------------+-------
+ ao_table_with_toast   |             0 |   160
+ ao_table_with_toast   |             1 | 50248
+ ao_table_with_toast   |             2 |   192
```
@red1452 , Test with and without patch shows the same size of segment files for AO tables. And  the toast relation with `reltoastrelid` has 0 size.  Why is it happening?

red1452 replied:
At this test, I add one row to toast table only at one segment. Empty toast table has zero size. Without patch sizes on segments will be the same as sizes before insert large value to this table.

bimboterminator1 replied:
Without and with patch the size of all segment files is the same (50248 at first segment) after inserting large value.
And the result of `adb_relation_storage_size` is also the same.  Toast table has 0 size. `calculate_toast_table_size` returns zero size.

bimboterminator1 replied:
Anyway, toast table has zero size for the column with large value. And test should show the case when toast table is filled eventually.

bimboterminator1 replied:
```
CREATE TABLE ao_table_with_toast(a INT, b TEXT)
WITH (APPENDOPTIMIZED=true, ORIENTATION=ROW);

insert into ao_table_with_toast select 1, (SELECT string_agg( chr(trunc(65+random()*26)::integer), '') FROM generate_series(1,50000));

SELECT oid::regclass AS Table,
       reltoastrelid::regclass As TOAST_Table,
       pg_relation_size(reltoastrelid) AS toast_size
FROM pg_class
WHERE relkind = 'r'
  AND reltoastrelid <> 0
ORDER BY 3 DESC; 
                  table                    |       toast_table       | toast_size 
--------------------------------------------+-------------------------+------------
ao_table_with_toast                        | pg_toast.pg_toast_16884 |      65536
```

bimboterminator1 replied:
**ORIENTATION=ROW** works

red1452 replied:
You are right, fixed.

red1452 committed:
fix test

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60281

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/879993

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/879994

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/879991

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/879988

red1452 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4476

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60475

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/889983

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/889984

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/889981

red1452 committed:
arenadata_toolkit: add toast's size to adb_relation_storage_size (#654)

This patch adds toast table's size to summary size of relation at C function
adb_relation_storage_size, which is used at SQL functions
adb_relation_storage_size and adb_relation_storage_size_on_segments.

The code is based on calculate_toast_table_size from dbsize.c, but without
calculating size of toast's indexes.

New test adb_relation_storage_size_test was added to check calculating sizes of
different tables (heap, ao, with toast or without) at SQL functions
adb_relation_storage_size and adb_relation_storage_size_on_segments.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/60488

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/890596

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/890597

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/890590
